Update readme to be more concise / code-forward

esDotDev opened the pull request:
@escamoteur Hey Thomas, I took a stab at making the readme a bit more concise / code-forward. 

Could you look it over and let me know what you think? 
* Have I missed anything important?
* Is anything wrong? 
* Do you think this is a good idea at all? 

[View readable](https://github.com/escamoteur/get_it_mixin/tree/update-readme)

Cheers,

esDotDev committed:
Initial stab at redoing readme

esDotDev committed:
Update readme

esDotDev committed:
Update README.md

esDotDev committed:
Update README.md

esDotDev committed:
Update README.md

esDotDev committed:
Update readme

escamoteur commented:
I just had a look at it. I like the idea to have a code example at the beginning but I'm not sure if it gets easier by making it that succinct. IMHO the original explains the concepts better in detail. We have to keep in mind that beginners too should be able to understand the usage.

esDotDev commented:
Hm, my thought is that this would generally be better for beginners to understand because it's more code-forward, and uses some tables to really clearly explain the various APIs. As a beginner, I generally just want to know how to _use_ a thing, and am not going to initially care about the how or the why until later.

The original readme seems more like a blog post that might accompany the launch of the lib, but I really think README's work best when they get down to business and just show show how to use the lib and explain the API.

esDotDev commented:
I personally had a quite hard time understanding all the API and the difference between `watchX` `watch` `watchOnly` etc based on the readme, and it wasn't until I found the doc page, and could just read the API's in a concise way, that I really understood the lib: https://pub.dev/documentation/get_it_mixin/latest/get_it_mixin/GetItMixin-mixin.html

Do you think this version can be augmented to be something you're happy with, or should we just scrap it?

Or, asked another way, is there any specific thing you think is missing, and maybe we can just add it in? I tried to copy from the original text wherever it made sense.

esDotDev committed:
Update README.md

esDotDev committed:
Update README.md

esDotDev committed:
Update README.md

escamoteur commented on README.md:
```diff
@@ -1,317 +1,177 @@
 
-This package offers a set of mixin types that makes the binding of data that is stored within `
-GetIt` really easy.
+A set of mixins that allow you to easily bind widgets to data registered with `GetIt`.
```
I would add 'these widgets will update automatically whenever the bound data changes.

escamoteur commented on README.md:
````diff
@@ -1,317 +1,177 @@
-class TestStateLessWidget1 extends StatelessWidget with GetItMixin {
+Here we bind to three `ValueListenable` which would normally be three builders, many lines of code and several levels of indentation. With `GetItMixin`, it's three lines:
+```dart
+class MyWidget extends StatelessWidget with GetItMixin {
````
I would add a snipped of the watched class type because watchOnly works differently

esDotDev replied:
Could you toss this in? I actually never use watchXOnly, or streams/futures... so probably going to mess something up :D

escamoteur commented on README.md:
````diff
@@ -1,317 +1,177 @@
-```
-or with the handler:
+There are some important rules to follow in order to avoid bugs with the `watch` methods:
+* `watch` methods should only be called within `build()`
````
should is too week. they don't work if you use them elsewhere
Also I would add "at the beginning of the build method

escamoteur commented on README.md:
```diff
@@ -1,317 +1,177 @@
+* `watch` methods should only be called within `build()`
+* must be called on every build, in the same order (no conditional watching)
+* do not use them inside of a builder as it will break the mixins ability to rebuild
 
```
It might be possible to call a register function in another place like 'initState' but I'm not sure. better not to give them ideas

esDotDev replied:
Ya I was going to ask you about this cause I don't know... it's not mentioned in the current docs. Seems like it would be a nice use case...registering each tick seems odd to me if you're in a stateful widget already.

escamoteur commented on README.md:
```diff
@@ -1,317 +1,177 @@
+|---|---|
+| `.registerHandler`  | Add a event handler for a `ValueListenable`  |
+| `.registerStreamHandler`  | Add a event handler for a `Stream`  |
+| `.registerStreamHandler`  | Add a event handler for a `Future`  |
```
I guess this should be registerFutureHandler :-)

esDotDev replied:
lol, ty!

escamoteur commented on README.md:
```diff
@@ -1,317 +1,177 @@
+    final currentUser = watchStream((UserModel x) => x.userNameUpdates, 'NoUser');
+    final ready = watchFuture((AppModel x) => x.initializationReady, false).data;
+    bool appLoadedAndUserLoggedIn = ready == true && currentUser.hasData;
+    ...
```
I would keep the columns of the original here because it puts the watch functions into a context

esDotDev replied:
Cool, but I don't think we need the column. How's this? 
```dart
    final currentUser = watchStream((UserModel x) => x.userNameUpdates, 'NoUser');
    final ready = watchFuture((AppModel x) => x.initializationReady, false).data;
    bool appIsLoading = ready == false || currentUser.hasData == false;
    
    if(appIsLoading) return CircularProgressIndicator();
    return Text(currentUser.data);    
```

escamoteur reviewed (changes requested):
Hi Shawn,
I have to apologize, I had only read the diff view of your changes which don't give a good impression of how it looks as a whole. 
I really like it. What do you think that we keep the original one as readme_extended and we add a link under the first example telling that if you like more information they can find a more prosa explanation here. We should also add a link to my talk https://youtu.be/YJ52kSfSMyM

escamoteur commented:
just as an addition, have you read https://github.com/fluttercommunity/get_it/issues/246

esDotDev commented:
Thanks

> Hi Shawn, I have to apologize, I had only read the diff view of your changes which don't give a good impression of how it looks as a whole. I really like it. What do you think that we keep the original one as readme_extended and we add a link under the first example telling that if you like more information they can find a more prosa explanation here. We should also add a link to my talk https://youtu.be/YJ52kSfSMyM

Nice! I'll review your comments and make the updates, and yes I think it's a good idea to do the links to the extended readme, as well as your talk.

I hadn't seen that link yet, looks like a good read, I'll dig in later today :)

esDotDev committed:
Update README.md

esDotDev committed:
Update README.md

esDotDev committed:
Update README.md

esDotDev committed:
Update README.md

esDotDev committed:
Update README.md

esDotDev committed:
Update README.md

esDotDev committed:
Update README.md

esDotDev commented:
@escamoteur I implemented most of your changes, and also updated the readme to use the phrasing 'watch' instead of 'bind' in some cases.

Remaining TBD: 
* Add an example data class for the watchXOnly example (maybe Streams/Futures as well?)
* Test whether we can registerHandler in initState, and clarify the rules if we can't
* Incorporate links to extended readme and video
 
Please have another look when you get a chance: https://github.com/escamoteur/get_it_mixin/tree/update-readme

esDotDev committed:
Update README.md

dancamdev commented:
Hi @escamoteur and @esDotDev,

I must say I like the proposed README quite a lot, it is more concise and offers a quick way to have a glance on what get_it_mixin is all about. Beginners would benefit a lot of the smaller, more concrete examples as well.

esDotDev commented:
@escamoteur Think we could finish this off soon? 

Here is what is left to tweak:
* Add an example data class for the watchXOnly example (maybe Streams/Futures as well?)
* Test whether we can `registerHandler` in `initState`, and clarify the rules if we can't
* Incorporate links to extended readme and video

escamoteur commented:
Hi Shawn, sorry for me not reacting, I hope it will go better for the next time. What do you watn from me to do here? I already agreed that we go forward with your readme witha link to the old one for people who like a more detailed explanation

esDotDev commented:
Hey! You mentioned "> I would add a snipped of the watched class type because watchOnly works differently", so I was hoping you could add that, as I'm not totally sure how it works. 

Also, I was curious:
* Can we call registerHandler in initState
* Do you have URLs we can use for the old readme? Should we just make it like history.md or something?

esDotDev committed:
Update README

esDotDev committed:
Update the readme

esDotDev commented:
Hi @escamoteur, I went ahead and added a link to the README_EXTENDED as well as your talk.

Think we are good to land this? Sorry for the looooong delay!
